sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/ExternalCatalogUtils.scala

```diff
@@ -63,22 +63,26 @@ object ExternalCatalogUtils {
     bitSet
   }
 
-  def needsEscaping(c: Char): Boolean = {
+  @inline private final def needsEscaping(c: Char): Boolean = {
```

Comment:
I wonder whether something like

```scala
  def escapePathName(path: String): String = {
    var builder: StringBuilder = null
    var i = 0
    while (i < path.length) {
      val c = path.charAt(i)
      if (c < charToEscape.size() && charToEscape.get(c)) {
        // Character needs escaping:
        if (builder eq null) {
          builder = new StringBuilder(path.substring(0, i))
          builder.append(path, 0, i)
        }
        builder.append('%')
        builder.append(f"${c.asInstanceOf[Int]}%02X")
      } else {
        // Character does not need escaping:
        builder.append(c)
      }
      i += 1
    }
    if (builder eq null) {
      path
    } else {
      builder.toString()
    }
  }
```

could be faster since it would avoid StringOps function calls for `indexWhere` plus function calls in `.foreach`. I think the cost of the builder null checks in the loop should be pretty cheap, although I can also appreciate how those checks might potentially impact loop unrolling. I think my suggestion should wind up being faster than what you have, even if it's not the theoretically fastest possible solution.

Reply:
Hi @JoshRosen I didn't get a positive benchmark result with your version on my local labtop

```
================================================================================================
Escape
================================================================================================

OpenJDK 64-Bit Server VM 17.0.10+0 on Mac OS X 14.5
Apple M2 Max
Escape Tests:                             Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Legacy                                             6085           6214         118          0.2        6084.6       1.0X
Kent                                               4849           4999         106          0.2        4849.2       1.3X
Josh                                               5536           5662          67          0.2        5536.0       1.1X
```

The benchmark I ran was
```scala
package org.apache.spark.sql

import org.apache.spark.benchmark.{Benchmark, BenchmarkBase}
import org.apache.spark.sql.catalyst.catalog.ExternalCatalogUtils

object EscapeBenchmark extends BenchmarkBase {
  override def runBenchmarkSuite(mainArgs: Array[String]): Unit = {
    val N = 1000000
    runBenchmark("Escape") {
      val benchmark = new Benchmark("Escape Tests", N, 10, output = output)
      val paths = Seq(
        "https://issues.apache.org/jira/browse/SPARK-48551",
        "https...issues.apache.org/jira/browse/SPARK-48551",
        "https...issues.apache.org.jira/browse/SPARK-48551",
        "https...issues.apache.org.jira.browse/SPARK-48551",
        "https...issues.apache.org.jira.browse.SPARK-48551")
      benchmark.addCase("Legacy") { _ =>
        (1 to N).foreach(_ => {
          paths.foreach(ExternalCatalogUtils.escapePathNameLegacy)
        })
      }

      benchmark.addCase("Kent") { _ =>
        (1 to N).foreach(_ => {
          paths.foreach(ExternalCatalogUtils.escapePathName)
        })
      }
      benchmark.addCase("Josh") { _ =>
        (1 to N).foreach(_ => {
          paths.foreach(ExternalCatalogUtils.escapePathName2)
        })
      }
      benchmark.run()
    }
  }
}
```

Reply:
FYI, I add diff below to avoid NPE.
```diff
  def escapePathName2(path: String): String = {
    var builder: StringBuilder = null
    var i = 0
    while (i < path.length) {
      val c = path.charAt(i)
      if (c < charToEscape.size() && charToEscape.get(c)) {
        // Character needs escaping:
        if (builder eq null) {
          builder = new StringBuilder(path.substring(0, i))
        }
        builder.append('%')
        builder.append(f"${c.asInstanceOf[Int]}%02X")
      } else {
+        if (builder eq null) {
+          builder = new StringBuilder()
+        }
        // Character does not need escaping:
        builder.append(c)
      }
      i += 1
    }
    if (builder eq null) {
      path
    } else {
      builder.toString()
    }
  }
```

I guess the problem with your code is that we can not take a detour into the original path sting, the StringBuilder cannot be avoided and each character is appended whether it will be used or not

Reply:
I think we could improve my original solution a bit by doing something like

```
} else if (builder ne null) {
  // Character does not need escaping but some earlier character did:
  builder.append(c)
}
```

but in further optimization I also found that the hex string conversion was expensive. After a bit more tinkering I ended up agreeing with your original approach of `indexWhere` since it avoids branching in the "escaping needed" loop.

Reply:
indexWhere and .foreach are replaced with 2 while loops